# 20210324

[x] edit R code to input all mutations file

I'm working on this, but it doesn't appear that the mutations are lining up correctly in the vcf file.

For example, when I compare the vcf output from Slim for the causal mutations to the 
VCF output from pyslim to the causal mutations, 
the mutations do not line up for the "same individual".

What it could be:

1 - it's a timing issue between when I output the trees file and output the vcf file 
(e.g. not the same generation or early vs. late) This happened to me before. HOWEVER,
In the slim file, `TOTAL_GEN` in simulations is 3000. 
The VCF file and tree sequences are both output at this time. UNLIKELY

1 - pyslim screws up the order of individuals from slim

I FOUND THE PROBLEM!

OK, so after I output the .vcf file from pyslim, I filter for MAF 0.01 with vcftools and recode.

When uploading into R, the SLIM vcf file has a "GT" in the "FORMAT" column of 
vcf@gt

```
tail(vcf@gt[keepmuts,1:10])
      FORMAT i0    i1    i2    i3    i4    i5    i6    i7    i8   
[42,] "GT"   "1|0" "0|0" "0|1" "0|0" "1|0" "0|0" "0|1" "0|0" "0|0"
[43,] "GT"   "0|0" "0|0" "0|0" "0|0" "0|0" "0|0" "0|0" "0|0" "0|0"
[44,] "GT"   "0|1" "1|1" "1|1" "1|1" "0|1" "1|1" "1|0" "1|1" "1|0"
[45,] "GT"   "0|0" "0|0" "0|0" "0|0" "0|0" "0|0" "0|0" "0|1" "0|0"
[46,] "GT"   "0|0" "0|0" "0|0" "0|0" "0|0" "0|0" "0|0" "0|0" "0|0"
[47,] "GT"   "0|0" "0|0" "0|0" "0|0" "0|0" "0|0" "0|0" "0|0" "0|0"
```

HOWEVER, the pyslim and vcftools filtered output has a genotype in the FORMAT column for CAUSAL MUTATIONS:

CAUSAL MUTATIONS:
```
 tail(vcf_full@gt[keepmuts_full,1:10])
      FORMAT tsk_0 tsk_1 tsk_2 tsk_3 tsk_4 tsk_5 tsk_6 tsk_7 tsk_8
[42,] "1|0"  "0|0" "0|1" "0|0" "1|0" "0|0" "0|1" "0|0" "0|0" "0|0"
[43,] "0|0"  "0|0" "0|0" "0|0" "0|0" "0|0" "0|0" "0|0" "0|0" "0|0"
[44,] "0|1"  "1|1" "1|1" "1|1" "0|1" "1|1" "1|0" "1|1" "1|0" "1|1"
[45,] "0|0"  "0|0" "0|0" "0|0" "0|0" "0|0" "0|0" "0|1" "0|0" "0|0"
[46,] "0|0"  "0|0" "0|0" "0|0" "0|0" "0|0" "0|0" "0|0" "0|0" "0|0"
[47,] "0|0"  "0|0" "0|0" "0|0" "0|0" "0|0" "0|0" "0|0" "0|0" "0|0"
```

BUT NOT FOR NON-CAUSAL mutations:
```
> head(vcf_full@gt[,1:10])
     FORMAT tsk_0 tsk_1 tsk_2 tsk_3 tsk_4 tsk_5 tsk_6 tsk_7 tsk_8
[1,] "GT"   "1|1" "1|1" "1|1" "1|1" "1|1" "1|1" "1|1" "1|1" "1|1"
[2,] "GT"   "1|1" "1|1" "1|1" "1|1" "1|1" "1|1" "1|1" "1|1" "1|1"
[3,] "GT"   "0|0" "0|0" "0|0" "0|0" "0|0" "0|0" "0|0" "0|0" "0|0"
[4,] "GT"   "0|0" "0|0" "0|0" "0|0" "0|0" "0|0" "0|0" "0|0" "0|0"
[5,] "GT"   "0|0" "1|1" "1|0" "1|1" "1|1" "0|1" "1|0" "1|0" "0|1"
[6,] "GT"   "0|0" "0|0" "0|0" "0|0" "0|0" "0|0" "0|0" "0|0" "0|0"
```

VIOLA! Found where they are not lining up. 

[] Go through code and make sure to edit so they line up!

* So, in BBEDit using grep I can find the issue locations with "\.\t\t"
* And I can replace them with "\.\t0\t"
* After doing the above in BBEdit, I was able to upload the VCF files into R and they loaded correctly
* Then out how to do that on the command line with sed
	* My mac doesn't like "\t"
	* My mac doesn't like the `-i` option in sed
	* `sed 's/\.		/\.	0	/g' ${outpath}${i}"_plusneut_MAF01.recode.vcf" > ${outpath}${i}"_plusneut_MAF01.recode2.vcf" #this works on my mac`
	




#### Plan:

[] Add LFMM to R code to quantify how often adaptive alleles are (1) not clinal or (2) clinal but not significant, or (3) clinal and outlier

	[] show that RDA gets messed up when adding complex environments
	[] add color to RDA plots to show the temp and salinity optimums of individuals - maybe use green to blue for salinity and light to dark for temperature
	[] ID outliers with LFMM - TEST IF RDA CAN PREDICT IND ENVIRONMENTS
	[] ID outliers with RDA - TEST IF RDA CAN PREDICT IND ENVIRONMENTS
	[] fix heat maps so the colors are constrained

[] set up a data frame of scenarios

[] Tweak unix pipeline to run a scenario quickly

[] test some of the other scenarios I want to create
