# 20210405

In the process of putting these outputs together, I realized that neither genotype-environment nor genotype-phenotype associations can discover the causal loci(!), even though they have additive effects on phenotypes, and the phenotypes are highly correlated with the environments. I'm digging in to make sure there isn't any bug, but it might happen because of the redundancy of the genetic architecture. There are a number of ways that mutations combine to produce non-extreme phenotypes, which reduces the association. Maybe plot heatmaps of phenotypes to understand this?

And yet, RDA can still predict relative salinity and temp phenotypes. so cool! The locus loadings to correspond to the phenotypic effects on salinity/temperature, but none of them are outliers by RDA either!

I wonder if power could be increased by aggregating allele frequencies within pops. I bet that makes a difference...

## Plan:

- [x] Add a neutral chromosome to the sims, that I can use to compare to results of outlier tests

- [x] Add LFMM to R code to quantify how often adaptive alleles are (1) not clinal or (2) clinal but not significant, or (3) clinal and outlier

- [ ] UPDATE LFMM TO LEA3

- [x]  ID outliers with LFMM - TEST IF RDA CAN PREDICT IND ENVIRONMENTS
- [x]  ID outliers with RDA - TEST IF RDA CAN PREDICT IND ENVIRONMENTS
    - From Capblanq 2018:
    "First, we recover the loci loadings from the RDA analysis. Only the loadings of the most informative ordination axis are kept for the rest of the procedure. The number of axes used (K) is determined by looking at the amount of information retained on the different axes of the RDA. A Mahalanobis distance D is then computed for each locus to identify loci showing extreme D values compared to the rest of the SNPs. A Mahalanobis distance is a multidimensional generalization of the idea of measuring how many standard deviations is a point from an average point. Computation of the Mahalanobis dis- tance uses the covRob function of the “ROBUST” R package (Wang et al., 2014). Mahalanobis distances are distributed as chi-squared distribu- tion with K degrees of freedom after correcting with the genomic infla- tion factor (Luu et al., 2017). Inflation factors are constant values that are used to rescale chi-square statistics in order to limit inflation due to diverse confounding factors (Fran"cois, Martins, Caye & Schoville, 2016). We then adjust the resulting p-values for the false discovery rate (FDR) by computing q-values with the “qvalue” R package (Storey, Bass, Dabney & Robinson, 2015). A locus is considered as an outlier if its q-value is <10%, meaning that 10% or less of the identified outliers could be false positives."

    - I used the function `rdadapt` from the 2018 paper supplement to do the calculation

- [ ] formalize all outputs
- [ ] add color to RDA plots to show the temp and salinity optimums of individuals - maybe use green to blue for salinity and light to dark for temperature
- [ ] fix heat maps so the colors are constrained
- [ ] test an obvious case that should produce clines
- [ ] set up a data frame of scenarios

- [ ] Tweak unix pipeline to run a scenario quickly


## Output tables

1. Use full VCF file
2. Calculate proportion of causal and neutral loci that show sig. correlations with temp (output correlation and slope)
  
#### Table 1: mutations

Done: 
  *`mut_ID` mutation ID, if not equal to 1, a causal mutation
  * `seed` simulation seed
  * `VCFrow` row in VCF file
  *`pos_pyslim` position as output from pyslim, which is 1 less than SliM output
  *`a_freq_full` allele frequency of derived allele based on all samples
  *`a_freq_subset` allele frequency of derived allele based on subset of individuals sampled according to their fitness
  *`mutSalEffect` for causal mutations, effect of mutation on the salinity phenotype
  *`mutTempEffect`  for causal mutations, effect of mutation on the temperature phenotype
  *`INFO`  for causal mutations, information output from slim
  *`af_cor_temp` correlation between allele frequency and temperature based on subset of individuals sampled according to their fitness
  *`af_cor_sal` correlation between allele frequency and salinity based on subset of individuals sampled according to their fitness
  *`af_slope_temp` slope between allele frequency and temperature based on subset of individuals sampled according to their fitness
  *`af_slope_sal` slope between allele frequency and salinity based on subset of individuals sampled according to their fitness
  * `Va_temp` VA to temperature explained by locus in metapopultion (this is a little misleading because some loci are low here, but explain a lot of VA in a specific environment), based on a subset of individuals after sampling based on their fitness
  * `Va_temp_prop` Proportion of total VA to temperature explained by locus in metapopultion (this is a little misleading because some loci are low here, but explain a lot of VA in a specific environment), based on a subset of individuals after sampling based on their fitness
  * `Va_sal` VA to salinity explained by locus in metapopultion (this is a little misleading because some loci are low here, but explain a lot of VA in a specific environment), based on a subset of individuals after sampling based on their fitness
  * `Va_sal_prop` Proportion of total VA to salinity explained by locus in metapopultion (this is a little misleading because some loci are low here, but explain a lot of VA in a specific environment), based on a subset of individuals after sampling based on their fitness
  
  Working on:
  
  (lots of this is output )
  * `LFMM_vXX_p` p-value from LFMM version XXX
  *`RDA_p` p-value from RDA in vegan as explained by Capblanq
  
  
 

  
  #### Table 2: summary of sim
  *`seed` 
  * `all_corr_phen_temp` for all individuals, correlation between individual temp phenotype and environment temperature
  * `subsamp_corr_phen_temp` after sampling 20 individuals from each patch with a probability based on their fitness, correlation individual temp phenotype and environment temperature
  * `all_corr_phen_sal` for all individuals, correlation between individual sal phenotype and environment salinity
  * `subsamp_corr_phen_sal` after sampling 20 individuals from each patch with a probability based on their fitness, correlation between individual sal phenotype and environment salinity
  *`num_causal` number of causal loci in sim
  *`num_neut` number of neutral loci in sim
  *`prop_causal_sig_temp_corr` proportion of causal loci that have significant correlations with temperature
  *`prop_causal_sig_sal_corr` proportion of causal loci that have significant correlations with salinity
  *`prop_neut_sig_temp_corr` proportion of neutral loci that have significant correlations with temperature
  *`prop_neut_sig_sal_corr` proportion of neutral loci that have significant correlations with salinity
  *`median_causal_temp_slope` median slope between allele frequency and temperature for causal loci
  *`median_causal_sal_slope` median slope between allele frequency and salinity for causal loci
  *`median_neut_temp_slope` median slope between allele frequency and temperature for neutral loci
  *`median_neut_sal_slope` median slope between allele frequency and salinity for neutral loci
  *`prop_causal_sig_LFMM` proportion of causal loci that are significant by LFMM
  *`prop_causal_sig_RDA` proportion of causal loci that are significant by RDA
  *`prop_VA_outliers_LFMM`  proportion of Va explained by LFMM outliers
  *`prop_VA_outliers_RDA` proportion of Va explained by RDA outliers
  *`prop_VA_outliers_RDA_LFMM` proportion of Va explained by both outliers from methods
  *`cor_RDAcausal_RDloadings_tempPhen` For an RDA based on causal loci, the correlation between (a linear prediction of the weighted RDA loadings) and (the individual's temp phenotype)
  *`cor_RDAcausal_RDloadings_salPhen` For an RDA based on causal loci, the correlation between (a linear prediction of the weighted RDA loadings) and (the individual's sal phenotype)
  *`cor_RDAalloutliers_RDloadings_tempPhen` For an RDA based on all outlier loci, the correlation between (a linear prediction of the weighted RDA loadings) and (the individual's temp phenotype)
  *`cor_RDAalloutliers_RDloadings_salPhen` For an RDA based on all outlier loci, the correlation between (a linear prediction of the weighted RDA loadings) and (the individual's sal phenotype)
  *`cor_RDAallSNPs_RDloadings_tempPhen` For an RDA based on all SNP (neutral and causal) loci, the correlation between (a linear prediction of the weighted RDA loadings) and (the individual's temp phenotype)
  *`cor_RDAallSNPs_RDloadings_salPhen` For an RDA based on all SNP (neutral and causal) loci, the correlation between (a linear prediction of the weighted RDA loadings) and (the individual's sal phenotype)


The following I decided not to output for now. It will take some time to calculate, and I'm not sure how I would present it.
* `prop_VaTemp_temp=1_sal=1` 
* `prop_VaTemp_temp=1_sal=-1`
* `prop_VaTemp_temp=-1_sal=-1`
* `prop_VaTemp_temp=-1_sal=1`
* `prop_VaTemp_temp=0_sal=-1`
* `prop_VaTemp_temp=0_sal=1`
* `prop_VaTemp_temp=0_sal=0`
* `prop_VaSal_temp=1_sal=1` 
* `prop_VaSal_temp=1_sal=-1`
* `prop_VaSal_temp=-1_sal=-1`
* `prop_VaSal_temp=-1_sal=1`
* `prop_VaSal_temp=0_sal=-1`
* `prop_VaSal_temp=0_sal=1`
* `prop_VaSal_temp=0_sal=0`
