



After a long hiatus of getting other papers out, I'm back!

Today I went through the rest of the `FinalAnalysis.md` R code, and made some more graphs for the main paper.

Next steps to do is to tweak a couple of things in the R code that I use to process the sims, and check a few outputs.

The main thing left is to implement the structure correction for RDA, and a few tweaks to make sure that I have the correct. 


### Minor tweaks
- [x] check FDR calculation is (FP/(TP + FP)) (correct for LFMM)
- [x] use the LG group and the position to name loci. And since mutations can occur on these same positions in the sims, I add a number so that my rownames are unique.
  - [x]  I noticed a place in the code where I had sorted by position instead of by the order in the VCF file; this may have affected some sims with duplicate positions. This is now fixed.
- [x] fix LG column of muts output. I noticed that one of the new muts files has some values in the LG column as missing, when this wasnâ€™t the case in the first file you gave me.
first file = /work/lotterhos/MVP-Offsets/practice_slim/mypractice/1231094_Rout_muts_full.txt
new muts = /work/lotterhos/MVP-Offsets/mypractice_20210308/slimdir/1231096_Rout_muts_full.txt

- [x] check "neutral" status of SNPs - sometimes I just use the `temp` column, this is correct because neutral SNPs are > position 500,000

### Fixes to SLiM code
- [x] The SLiM code somehow got increased to 21 chromosomes - it should be 20 (10 with causal and 10 neutral) -  I fixed it back to 20, but should rerun the sims 
- [x] think about N-equal, m-breaks scenario. not really giving different FST or K, difference in structure only slight
  - [x] old `mig_breaks_SS = 0.01;` ; new `mig_breaks_SS = 0.005`;
  - [x] old `mig_breaks_estuary = 0.03`; new `mig_breaks_estuary = 0.015;`


## Investiaging bad RDA predictions
* Simulation 1231144 - all the figures look good, like the prediction should be good. But the prediction was poor. Same for other figures. I will need to double check this, but it appears that all the problematic cases have two traits. Some have plieotropy and some don't. With two traits, both RDA loadings contribute a lot to the calculation. 

### RDA
- [x] RDA code to use https://github.com/Capblancq/RDA-landscape-genomics
  - [x] Note that in Thibaut's application he uses allele frequencies, while I use the raw genotypes.
- [x] implement structure correction for RDA
  - At first I tried adding geography (x,y location) and structure (PC axis) to the RDA, this completely corrected for everything, which led to very low performance
  - To give RDA a fighting chance, I used just the first PCs of structure. This seemed to result is a slightly higher correlation with the individual scores.
- [x] Fixed some code with the scaling of the arrows in the RDA plotting
 
- [x] Why do some RDA predictions seem to fail?
  - [x] Started with 1231144, but 1232947 looks like an interesting case.
  - [x] After exploring this a bit more and editing the code, I couldn't find what the issue was. I edited a bit of code to make sure none of the variables had duplicate names, so it's possible there was a bug that I fixed and didn't realize it?
  - [x] There definitely was a bug in the mutation prediction, getting info from different RDA objects
  
- [x] RDA - compare the ancestral/derived signed approach (what was used in Laruson) to other signing approaches
  - first I tried just changing the sign of ~8000 out of ~25000 loci to see what would happen. This didn't seem to affect anything.


### Sync github

Put outputs from the last run into this folder: `mv sim_output_20220201 sim_output_20220201_20220420`

I moved all the other graphs into a folder: `archive/graphs_20220420/`

```
git status
git add <added files in commit stages>
git commit
git pull
git push

```


  
### ORDER OF OUTPUTS - ADD TO METADATA
```

out_full <- data.frame(seed=as.character(seed),
                       n_samp_tot=nrow(subset_indPhen_df), 
                       n_samp_per_pop = n_per_pop,
                       sd_fitness_among_inds,
                       sd_fitness_among_pops,
                       final_LA = LA_df$local_adapt[nrow(LA_df)],
                       K, Bonf_alpha,
                       numCausalLowMAFsample,
                       all_corr_phen_temp,
                       subsamp_corr_phen_temp,
                       all_corr_phen_sal,
                       subsamp_corr_phen_sal,
                       num_causal_prefilter,
                       num_causal_postfilter,
                       num_non_causal = sum(!muts_full$causal),
                       num_neut_prefilter = num_neut_prefilter, #total number non-causal loci
                       num_neut_postfilter = num_neut_prefilter, #total number non-causal loci after filtering
                       num_neut_neutralgenome = sum(muts_full$causal_temp=="neutral"), #loci in 2nd half of genome
                       num_causal_temp = sum(muts_full$causal_temp=="causal"),
                       num_causal_sal = sum(muts_full$causal_sal=="causal"),
                       num_multiallelic,
                       
                       meanFst,
                       va_temp_total,
                       va_sal_total,
                       
                       Va_temp_sample,
                       Va_sal_sample,
                       nSNPs, #` total number of SNPs in analysis
                      
                      median_causal_temp_cor, # median correlation between AF and Temp for causal loci
                       median_causal_sal_cor, 
                       median_neut_temp_cor, 
                       median_neut_sal_cor,           
                       
                # CORRELATION OUTPUTS       
                       cor_VA_temp_prop,
                       cor_VA_sal_prop,
                       cor_TPR_temp,
                       cor_TPR_sal,
                       
                       cor_FDR_allSNPs_temp,
                       cor_FDR_neutSNPs_temp,
                       cor_FDR_allSNPs_sal,
                       cor_FDR_neutSNPs_sal,
                      num_causal_sig_temp_corr, ## number of causal loci that have significant Kendall's correlations with temperature after Bonferroni correction
                      num_causal_sig_sal_corr, ## number of causal loci that have significant Kendall's correlations with salinity after Bonferroni correction
                      num_notCausal_sig_temp_corr, # number of non-causal (neutral and neutral-linked) loci that have significant Kendall's correlations with temperature after Bonferroni correction
                      num_notCausal_sig_sal_corr, # number of non-causal (neutral and neutral-linked) loci that have significant Kendall's correlations with salinity after Bonferroni correction
                      num_neut_sig_temp_corr, # number of neutral (unlinked to causal) loci that have significant Kendall's correlations with temperature after Bonferroni correction
                      num_neut_sig_sal_corr,# number of neutral (unlinked to causal) loci that have significant Kendall's correlations with salinity after Bonferroni correction
                             
                       cor_AUCPR_temp_allSNPs,
                       cor_AUCPR_temp_neutSNPs,
                       cor_AUCPR_sal_allSNPs,
                       cor_AUCPR_sal_neutSNPs,
                       cor_af_temp_noutliers, # number of outliers for cor(af,temp) after Bonferroni correction
                       cor_af_sal_noutliers, # number of outliers for cor(af,salinity) after Bonferroni correction
                       cor_FPR_temp_neutSNPs, #` false positive rate in cor(af,temp) after Bonferroni correction, based on loci unaffected by selection
                       cor_FPR_sal_neutSNPs, #` false positive rate in cor(af,sal) after Bonferroni correction, based on loci unaffected by selection
                      
                # LFMM OUTPUTS 
                      LEA3.2_lfmm2_Va_temp_prop, 
                      LEA3.2_lfmm2_Va_sal_prop, 
                      LEA3.2_lfmm2_TPR_temp,
                      LEA3.2_lfmm2_TPR_sal,

                      LEA3.2_lfmm2_FDR_allSNPs_temp, 
                      LEA3.2_lfmm2_FDR_allSNPs_sal, 
                      LEA3.2_lfmm2_FDR_neutSNPs_temp,
                      LEA3.2_lfmm2_FDR_neutSNPs_sal,
                      
                      LEA3.2_lfmm2_AUCPR_temp_allSNPs,
                      LEA3.2_lfmm2_AUCPR_temp_neutSNPs,
                      LEA3.2_lfmm2_AUCPR_sal_allSNPs,
                      LEA3.2_lfmm2_AUCPR_sal_neutSNPs,
                      LEA3.2_lfmm2_mlog10P_tempenv_noutliers, #` number of outliers for the lfmm temp model (qvalue <0.05)
                      LEA3.2_lfmm2_mlog10P_salenv_noutliers, #` number of outliers for the lfmm salinity model (qvalue <0.05)
                      LEA3.2_lfmm2_num_causal_sig_temp, #` number of causal loci on the temp trait, significant in the lfmm temp model (qvalue <0.05)
                      LEA3.2_lfmm2_num_neut_sig_temp, #` number of neutral loci false positives (only neutral loci not affected by selection), significant in the lfmm temp model (qvalue <0.05)
                      LEA3.2_lfmm2_num_causal_sig_sal, #` number of causal loci on the salinity trait, significant in the lfmm salinity model (qvalue <0.05)
                      LEA3.2_lfmm2_num_causal_sig_temp,
                      LEA3.2_lfmm2_FPR_neutSNPs_temp, #` false positive rate of lfmm temperature model
                      LEA3.2_lfmm2_FPR_neutSNPs_sal, #` false positive rate of lfmm salinity model
                
                #RDA outputs
                      RDA1_propvar, #proportion of variance explained by 1st RDA axis
                      RDA2_propvar, #proportion of variance explained by 2nd RDA axis
                      RDA1_propvar_corr, #proportion of variance explained by 1st RDA axis with structure correction
                      RDA2_propvar_corr, #proportion of variance explained by 2nd RDA axis with structure correction
                      
                      RDA1_temp_cor, #` output of `summary(rdaout)$biplot[2,1]`, which is the correlation between RDA1 and the temperature environmental variable
                      RDA1_sal_cor, #`  output of `summary(rdaout)$biplot[1,1]`, which is the correlation between RDA1 and the salinity environmental variable
                      RDA2_temp_cor, #` output of `summary(rdaout)$biplot[2,2]`, which is the correlation between RDA2 and the temperature environmental variable
                      RDA2_sal_cor, #` output of `summary(rdaout)$biplot[1,2]`, which is the correlation between RDA2 and the salinity environmental variable
                      
                # RDA OUTLIER OUTPUTS
                      RDA_Va_temp_prop,
                      RDA_Va_temp_prop_corr, #prop of VA explained by outliers with structure correction
                      RDA_Va_sal_prop,
                      RDA_Va_sal_prop_corr, #prop of VA explained by outliers with structure correction
                      RDA_TPR, 
                      RDA_TPR_corr, #with structure correction
                      RDA_FDR_allSNPs,
                      RDA_FDR_allSNPs_corr,#with structure correction
                      num_RDA_sig_causal,
                      num_RDA_sig_neutral, #only for neutral half of genome
                      num_RDA_sig_causal_corr, #with structure correction
                      num_RDA_sig_neutral_corr, #with structure correction
                      RDA_FDR_neutSNPs,
                      RDA_FDR_neutSNPs_corr,  #with structure correction
                      RDA_AUCPR_allSNPs,
                      RDA_AUCPR_neutSNPs,
                      RDA_AUCPR_neutSNPs_corr, #with structure correction
                      RDA_FPR_neutSNPs, 
                      RDA_FPR_neutSNPs_corr, #with structure correction
          
                # RDA PREDICTION OUTPUTS     
                      RDA_RDAmutpred_cor_tempEffect, #kendall's correlation between the predicted temperature effect from RDA and the true mutation effect on temperature
                      RDA_RDAmutpred_cor_salEffect, # kendall's correlation between the predicted salinity effect from RDA and the true mutation effect on salinity
                      RDA_absRDAmutpred_cor_tempVa, # kendall's correlation between the abs(predicted temperature effect from RDA) and the true mutation Va on temperature
                      RDA_absRDAmutpred_cor_salVa, #kendall's correlation between the abs(predicted salinity effect from RDA) and the true mutation Va on salinity
                    
                      RDA_RDAmutpred_cor_tempEffect_structcorr,
                      RDA_RDAmutpred_cor_salEffect_structcorr,
                      RDA_absRDAmutpred_cor_tempVa_structcorr,
                      RDA_absRDAmutpred_cor_salVa_structcorr,
                      
                      RDA_cor_RDA20000temppredict_tempPhen, 
                      RDA_cor_RDA20000salpredict_salPhen,
                      RDA_cor_RDA20000temppredict_tempPhen_structcorr,
                      RDA_cor_RDA20000salpredict_salPhen_structcorr,
                

            # STRUCUTRE CORRELATION OUTPUTS
                      cor_PC1_temp, 
                      cor_PC1_sal, 
                      cor_PC2_temp, 
                      cor_PC2_sal, 
                      cor_LFMMU1_temp, 
                      cor_LFMMU1_sal, 
                      cor_LFMMU2_temp, 
                      cor_LFMMU2_sal,
                      cor_PC1_LFMMU1_temp, 
                      cor_PC1_LFMMU1_sal, 
                      cor_PC2_LFMMU1_temp, 
                      cor_PC2_LFMMU1_sal
)
```
  
  ### ADD TO MUTS DF
```
  muts_full$RDA_mlog10P <- -log10(ps$p.values)
  muts_full$RDA_mlog10P_sig <- ps$q.values<0.05
  muts_full$RDA_mlog10P_corr <- -log10(ps_corr$p.values)
  muts_full$RDA_mlog10P_sig_corr <- ps_corr$q.values<0.05
    muts_full$RDA1_score <- loci.sc[,1]
  muts_full$RDA2_score <- loci.sc[,2]
  muts_full$RDA1_score_corr <- loci.sc_corr[,1]
  muts_full$RDA2_score_corr <- loci.sc_corr[,2]
  muts_full$RDA_mut_temp_pred_structcorr
  muts_full$RDA_mut_sal_pred_structcorr 
```

### ADD TO IND DF
```
  subset_indPhen_df$RDA1 <- ind.sc[,1]
  subset_indPhen_df$RDA2 <- ind.sc[,2]
  subset_indPhen_df$RDA1_corr <- ind.sc_corr[,1]
  subset_indPhen_df$RDA2_corr <- ind.sc_corr[,2]
  subset_indPhen_df$RDA_PC1 <-  ind.sc[,3]
  subset_indPhen_df$RDA_PC2 <-  ind.sc[,4]
  subset_indPhen_df$RDA_PC1_corr <-  ind.sc_corr[,3]
  subset_indPhen_df$RDA_PC2_corr <-  ind.sc_corr[,4]
  subset_indPhen_df$RDA_predict_tempPhen_20KSNPs <- temp_pred
  subset_indPhen_df$RDA_predict_salPhen_20KSNPs <- sal_pred
  subset_indPhen_df$RDA_predict_tempPhen_20KSNPs_structcorr <- temp_pred_corr
  subset_indPhen_df$RDA_predict_salPhen_20KSNPs_structcorr <- sal_pred_corr
```  
  
  
