# Results from previous post

## Failed sims

# Some sims (~50) still failed due to low number of SNPs after sampling

SNPs prior to sampling:
```
(base) [lotterhos@login-00 MVP-NonClinalAF]$ wc -l sim_output_20220201/1231163_plusneut_MAF01.recode2.vcf.gz
8481 sim_output_20220201/1231163_plusneut_MAF01.recode2.vcf.gz
```

## Review recombination rate
* defineConstant("R", 1e-6);
* Each LG 50000 sites, 20 LGs
* N = 100 * 100 = 10,000

* $N_er$ = 10,000 * 1e-06 = 0.01 which gives a resolution of 0.001 cM between proximate bases.
* Length: 0.001 cM * 50,000 sites = 50 cM

We simulated 21 linkage groups (LGs), each 10 cM long, with a resolution of 0.0001 cM between proximate bases in SLiM. 
In 20 of the LGs QTN mutations could arise, and the 21st was neutral. Thus, our simulations represent the case where unlinked locations 
affected by selection were tracked within a much larger genome. The recombination rate was scaled to mimic the case where SNPs were 
collected across each LG (similar to a SNP chip), but still low enough to allow signatures of selection to arise in neutral loci 
linked to selected loci (in the simulations 100,000 bases was 10 cM; in humans 100,000 bp would correspond to 0.1 cM). 


From Lotterhos 2019:

The base recombination rate Ner = 0.01 (unless manipulated as described below) gave a resolution of 0.001 cM between proximate bases. 
The recombination rate was scaled to mimic the case where SNPs were collected across a larger genetic map than what was simulated 
(similar to a SNP chip), but still low enough to allow signatures of selection to arise in neutral loci linked to selected loci 
(in the simulations 50,000 bases / (r = 1e-05) * 100 = 50 cM; in humans 50,000 bp would correspond to 0.05 cM). 
Thus, SNPs at the opposite ends of linkage groups were likely to have a recombination rate between them of 0.5 (unlinked), 
but there would otherwise be some degree of linkage among SNPs within linkage groups. 
For all LGs, the population-scaled mutation rate NeÎ¼ equaled 0.001.

## Inspect FST of simulations

![image](https://user-images.githubusercontent.com/6870125/153580268-01915a00-5c97-4a67-a675-51d2039cd3bd.png)

Wow, Fst for N-variable m-variable was all over the place!

![image](https://user-images.githubusercontent.com/6870125/153580544-6fb09a56-97eb-4364-9a29-24c6366a2385.png)

```
           Fst        LA
SS-Clines_N-equal_m-constant                 0.05782970 0.4841922
SS-Mtn_N-equal_m-constant                    0.06339836 0.4862318
Est-Clines_N-equal_m-constant                0.08777946 0.4297248
SS-Clines_N-cline-N-to-S_m-constant          0.12456721 0.4798241
SS-Mtn_N-cline-N-to-S_m-constant             0.13284979 0.4779571
Est-Clines_N-cline-N-to-S_m-constant         0.15853110 0.4226462
SS-Clines_N-cline-center-to-edge_m-constant  0.12918562 0.4781890
SS-Mtn_N-cline-center-to-edge_m-constant     0.13925447 0.4797316
Est-Clines_N-cline-center-to-edge_m-constant 0.18727872 0.4210846
SS-Clines_N-equal_m_breaks                   0.06927258 0.4886862
SS-Mtn_N-equal_m_breaks                      0.07520083 0.4905438
Est-Clines_N-equal_m_breaks                         NaN       NaN
SS-Clines_N-variable_m-variable              0.29169034 0.4693067
SS-Mtn_N-variable_m-variable                 0.32879766 0.4657067
Est-Clines_N-variable_m-variable                    NaN       NaN
```

# New SLiM simulations

This is really confusing, given the much higher FST, that the number of SNPs is still very small.

One way to reduce variation is to sample from fewer migration rates, and change the way N is sampled.


Old params: `var_m_ss = c(rep(0.001,3), rep(0.01,10),rep(0.1,5),rep(0.25,1));`

New params: `var_m_ss = c(rep(0.01,10), rep(0.1,5));` (mean 0.04)



I also edited the R code so it wouldn't break if there was low diversity in the simulation.

## Running new sims for N-var m-var

```
(base) [lotterhos@d3037 src]$ sbatch d-run_nonAF_sims_0Slim-fastruns-20220201-redo.sh
Submitted batch job 23195110
```

<img width="837" alt="image" src="https://user-images.githubusercontent.com/6870125/153584207-633f3fbf-2d79-421f-ac7d-9877e499f942.png">

## Sim results

`ls -l *_Rout_simSummary.txt | wc -l # this is the number of sims that were analyzed through the final output in the R script`


That's good!

awk 'NR==1' 1231094_Rout_simSummary.txt > ../summary_20220201.txt # header

awk 'FNR>1' *_Rout_simSummary.txt >> ../summary_20220201.txt # data


```
                                                Fst        LA
SS-Clines_N-equal_m-constant                 0.05782970 0.4841922
SS-Mtn_N-equal_m-constant                    0.06339836 0.4862318
Est-Clines_N-equal_m-constant                0.08777946 0.4297248
SS-Clines_N-cline-N-to-S_m-constant          0.12456721 0.4798241
SS-Mtn_N-cline-N-to-S_m-constant             0.13284979 0.4779571
Est-Clines_N-cline-N-to-S_m-constant         0.15853110 0.4226462
SS-Clines_N-cline-center-to-edge_m-constant  0.12918562 0.4781890
SS-Mtn_N-cline-center-to-edge_m-constant     0.13925447 0.4797316
Est-Clines_N-cline-center-to-edge_m-constant 0.18727872 0.4210846
SS-Clines_N-equal_m_breaks                   0.06927258 0.4886862
SS-Mtn_N-equal_m_breaks                      0.07520083 0.4905438
Est-Clines_N-equal_m_breaks                         NaN       NaN
SS-Clines_N-variable_m-variable              0.25515287 0.4752973
SS-Mtn_N-variable_m-variable                 0.26449298 0.4728338
Est-Clines_N-variable_m-variable                    NaN       NaN
```

![image](https://user-images.githubusercontent.com/6870125/153709034-8657c000-bd5b-43e5-9d37-b703ba3d8e9f.png)


# Running new sims for N-var m-var

Old params: `var_m_ss = c(rep(0.01,10), rep(0.1,5));` (mean m ~ 0.04, gives an FST of 0.25)

New params: `var_m_ss = c(rep(0.01,10), rep(0.1,5), rep (0.25, 5);` 

I revised the code so that equal numbers of high and low N demes would be chosen:
`N_var = sample(rep(N,10), 100, replace=FALSE)`
`sim.addSubpop(i, N_var[i-1]); // variable N`

```
(base) [lotterhos@login-00 src]$ sbatch d-run_nonAF_sims_0Slim-fastruns-20220201-redo.sh
Submitted batch job 23196870
```

```
> cbind(Fst=a,LA=b)
                                                    Fst        LA
SS-Clines_N-equal_m-constant                 0.05782970 0.4841922
SS-Mtn_N-equal_m-constant                    0.06339836 0.4862318
Est-Clines_N-equal_m-constant                0.08777946 0.4297248
SS-Clines_N-cline-N-to-S_m-constant          0.12456721 0.4798241
SS-Mtn_N-cline-N-to-S_m-constant             0.13284979 0.4779571
Est-Clines_N-cline-N-to-S_m-constant         0.15853110 0.4226462
SS-Clines_N-cline-center-to-edge_m-constant  0.12918562 0.4781890
SS-Mtn_N-cline-center-to-edge_m-constant     0.13925447 0.4797316
Est-Clines_N-cline-center-to-edge_m-constant 0.18727872 0.4210846
SS-Clines_N-equal_m_breaks                   0.06927258 0.4886862
SS-Mtn_N-equal_m_breaks                      0.07520083 0.4905438
Est-Clines_N-equal_m_breaks                         NaN       NaN
SS-Clines_N-variable_m-variable              0.21002911 0.4509264
SS-Mtn_N-variable_m-variable                 0.21777354 0.4348857
Est-Clines_N-variable_m-variable                    NaN       NaN
```

![image](https://user-images.githubusercontent.com/6870125/153721973-aec078f6-50ba-4993-9dc2-9d975465d4fe.png)


## Determine sims for Est-Clines_N-equal_m_breaks and Est-Clines_N-variable_m-variable 

```
(base) [lotterhos@login-00 src]$ sbatch d-run_nonAF_sims_0Slim-longruns-20220201.sh
Submitted batch job 23200523
```
