## `seed_Rout_simSummary.txt` (output after analysis of simulations) A table with information about the entire simulation

After each simulation is analyzed in R, the following summary statistics are output.
This file is best managed when merged with `0b-final_params-20220428.txt`

Correlations are based on Kendall's rank correlation unless otherwise stated.

`Salinity` was revised to the more general `Env2` for the main manuscript

-------------------------
### Basic outputs
-------------------------

*`seed` simulation seed
* `n_samp_tot` total number of individuals sampled
* `nind_samp_per_pop` number of individuals sampled from each deme
* `sd_fitness_among_inds` variance in fitness among all sampled individuals in the simulation (sampling prob. is proportional to fitness to mimic viability selection)
* `sd_fitness_among_pops` variance in fitness among all demes in the simulation after sampling (sampling prob. is proportional to fitness to mimic viability selection)
* `final_LA` final amount of local adaptation in the simulation
* `K` number of populations used in analyses
* `Bonf_alpha` the significance threshold for P-values applied to the correlation
* `numCausalLowMAFsample` number of causal loci that were not filtered out, but were below the MAF cutoff. These were included in the calculations. **REVISIT**
* `all_corr_phen_temp` for all individuals, correlation between individual temp phenotype and environment temperature
* `subsamp_corr_phen_temp` after sampling 10 individuals from each deme with a probability based on their fitness, correlation individual temp phenotype and environment temperature
* `all_corr_phen_sal` for all individuals, correlation between individual sal phenotype and environment salinity
* `subsamp_corr_phen_sal` after sampling 10 individuals from each deme with a probability based on their fitness, correlation between individual sal phenotype and environment salinity

* `num_causal_prefilter` number of causal loci in sim before filtering for MAF > 0.01
* `num_causal_postfilter` number of causal loci in sim after filtering for MAF > 0.01
* `num_non_causal` number of neutral loci in sim arising on the half of the genome where they could be linked to causal loci
* `num_neut_prefilter` total number of neutral loci on all LG before filtering MAF > 0.01. This is not really accurate, since many neutral loci were filtered out after output by pyslim - causal loci were not subject to filtering.
* `num_neut_postfilter`  total number of neutral loci on all LG after filtering MAF > 0.01
* `num_neut_neutralgenome` number of truly neutral loci in sim, unlinked to causal loci, on LG 11-20
* `num_causal_temp` number of loci with non-zero phenotypic effects on the temperature phenotype
* `num_causal_sal`number of loci with non-zero phenotypic effects on the salinity phenotype
* `num_multiallelic` Rarely there is a back-mutation in SLiM, leading to a 0/0 0/1 1/1 1/2 0/2 2/2 genotypes. These were filtered for analysis.

* `mean_Fst` overall FST calculated from mean(T1)/mean(T2) in outflank
* `va_temp_total` total additive genetic variance in the temperatuer trait, based on the entire 10,000 individual sample
* `va_sal_total` total additive genetic variance in the salinity trait, based on the entire 10,000 individual sample
* `Va_temp_sample` total additive genetic variance in the temperatuer trait, based on the 1,000 individual (10 ind/deme x 100 demes) sample
* `Va_sal_sample` total additive genetic variance in the salinity trait, based on the 1,000 individual (10 ind/deme x 100 demes) sample
* `nSNPs` total number of SNPs in analysis

-------------------------
### Outputs based on Kendall's correlation between genotype and environment
-------------------------

* `median_causal_temp_cor` median abs(Spearman's correlation) between allele frequency and temperature for causal loci
* `median_causal_sal_cor` median abs(Spearman's correlation)  between allele frequency and salinity for causal loci
* `median_neut_temp_cor` median abs(Spearman's correlation)  between allele frequency and temperature for neutral loci
* `median_neut_sal_cor` median abs(Spearman's correlation)  between allele frequency and salinity for neutral loci

* `cor_VA_temp_prop` proportion of VA in temperature phenotype explained by clinal outliers for temperature, based on kendall's correlation between deme allele frequency and deme temperature
* `cor_VA_sal_prop`  proportion of VA in salinity phenotype explained by clinal outliers for salinity, based on kendall's correlation between deme allele frequency and deme salinity
* `cor_TPR_temp` true positive rate for loci with non-zero effects on temperature, based on kendall's correlation between deme allele frequency and deme temperature 
* `cor_TPR_sal`true positive rate for loci with non-zero effects on salinity, based on kendall's correlation between deme allele frequency and deme salinity

* `cor_FDR_allSNPs_temp` false discovery rate of (kendall's correlation between deme allele frequency and deme temperature) for loci with non-zero effects on temperature 
* `cor_FDR_neutSNPs_temp` an optimistic calculation for false discovery rate of  (kendall's correlation between deme allele frequency and deme temperature) for loci with non-zero effects on temperature, excluding non-causal loci in half of genome affected by selection 
* `cor_FDR_allSNPs_sal` false discovery rate of (kendall's correlation between deme allele frequency and deme salinity) for loci with non-zero effects on salinity
* `cor_FDR_neutSNPs_sal` an optimistic calculation for false discovery rate of  (kendall's correlation between deme allele frequency and deme temperature) for loci with non-zero effects on temperature, excluding non-causal loci in half of genome affected by selection

* `num_causal_sig_temp_corr` Number of causal loci on temperature trait that are significant cor(af,temp) after Bonferroni correction
* `num_causal_sig_sal_corr` Number of causal loci on salinity trait that are significant cor(af,salinity) after Bonferroni correction
* `num_notCausal_sig_temp_corr` Number of non-causal (neutral and neutral-linked) loci that are significant cor(af,temp) after Bonferroni correction
* `num_notCausal_sig_sal_corr` Number of non-causal (neutral and neutral-linked) loci that are significant cor(af,salinity) after Bonferroni correction
* `num_neut_sig_temp_corr` Number of truly neutral loci (LG 11-20) that are significant cor(af,temp) after Bonferroni correction
* `num_neut_sig_sal_corr` Number of truly neutral loci (LG 11-20) that are significant cor(af,salinity) after Bonferroni correction

* `cor_AUCPR_temp_allSNPs` AUC-PR of kendall's correlation between deme allele frequency and deme temperature, based on the whole genome and causal loci for temperature
* `cor_AUCPR_temp_neutSNPs` an optimistic estimate of AUC-PR of kendall's correlation between deme allele frequency and deme temperature, based on causal loci for temperature and neutral loci not affected by selection (excluding non-causal loci in half of genome affected by selection)
* `cor_AUCPR_sal_allSNPs` AUC-PR of kendall's correlation between deme allele frequency and deme salinity, based on the whole genome and causal loci for salinity
* `cor_AUCPR_sal_neutSNPs` an optimistic estimate of AUC-PR of kendall's correlation between deme allele frequency and deme salinity, based on causal loci for salinity and neutral loci not affected by selection (excluding non-causal loci in half of genome affected by selection)

* `cor_af_temp_noutliers` number of outliers for cor(af,temp) after Bonferroni correction
* `cor_af_sal_noutliers` number of outliers for cor(af,salinity) after Bonferroni correction
* `cor_FPR_temp_neutSNPs` false positive rate in cor(af,temp) after Bonferroni correction, based on neutral loci unaffected by selection (LG 11-20)
* `cor_FPR_sal_neutSNPs` false positive rate in cor(af,sal) after Bonferroni correction, based on  neutral loci unaffected by selection (LG 11-20)


-------------------------
### LFMM outputs 
-------------------------

LFMM significance was based on q-value < 0.05

* `LEA3.2_lfmm2_Va_temp_prop` proportion of additive genetic variance (Va) in the temperature trait explained by outliers in the LFMM temp model
* `LEA3.2_lfmm2_Va_sal_prop` proportion of additive genetic variance (Va) in the saliniity trait explained by outliers in the LFMM salinity model

*  `LEA3.2_lfmm2_TPR_temp` true positive rate of the LFMM temp model for loci with alleles that have non-zero effects on the temperature phenotype
*  `LEA3.2_lfmm2_TPR_sal` true positive rate of the LFMM salinity model for loci with alleles that have non-zero effects on the salnity phenotype

* `LEA3.2_lfmm2_FDR_allSNPs_temp` false discovery rate of the LFMM temp model for the entire genome      
* `LEA3.2_lfmm2_FDR_allSNPs_sal` false discovery rate of the LFMM temp model for the entire genome   
* `LEA3.2_lfmm2_FDR_neutSNPs_temp` an optimistic calculation of the false discovery rate of the LFMM temp model, including only causal loci and neutral loci not affected by selection (any non-causal loci that arises on the half of the genome affected by selection was excluded)
* `LEA3.2_lfmm2_FDR_neutSNPs_sal` an optimistic calculation of the false discovery rate of the LFMM salinity model, including only causal loci and neutral loci not affected by selection (any non-causal loci that arises on the half of the genome affected by selection was excluded)

* `LEA3.2_lfmm2_AUCPR_temp_allSNPs` the AUC-PR of the lfmm temp model based on the entire genome  
* `LEA3.2_lfmm2_AUCPR_temp_neutSNPs` an optimistic calculation of the AUC-PR of the LFMM temp model, including only causal loci and neutral loci not affected by selection (any non-causal loci that arises on the half of the genome affected by selection was excluded)
* `LEA3.2_lfmm2_AUCPR_sal_allSNPs` the AUC-PR of the lfmm salinity model based on the entire genome
* `LEA3.2_lfmm2_AUCPR_sal_neutSNPs` an optimistic calculation of the AUC-PR of the LFMM salinity model, including only causal loci and neutral loci not affected by selection (any non-causal loci that arises on the half of the genome affected by selection was excluded) 

* `LEA3.2_lfmm2_mlog10P_tempenv_noutliers` number of outliers for the lfmm temp model (qvalue <0.05)
* `LEA3.2_lfmm2_mlog10P_salenv_noutliers` number of outliers for the lfmm salinity model (qvalue <0.05)

* `LEA3.2_lfmm2_num_causal_sig_temp` number of causal loci on the temp trait, significant in the lfmm temp model (qvalue <0.05)
* `LEA3.2_lfmm2_num_neut_sig_temp` number of neutral loci false positives (only neutral loci not affected by selection), significant in the lfmm temp model (qvalue <0.05)
* `LEA3.2_lfmm2_num_causal_sig_sal` number of causal loci on the salinity trait, significant in the lfmm salinity model (qvalue <0.05)
* `LEA3.2_lfmm2_num_neut_sig_sal` number of neutral loci false positives (only neutral loci not affected by selection), significant in the lfmm salinity model (qvalue <0.05)

*  `LEA3.2_lfmm2_FPR_neutSNPs_temp` false positive rate of lfmm temperature model
*  `LEA3.2_lfmm2_FPR_neutSNPs_sal` false positive rate of lfmm salinity model

-------------------------
### RDA outputs 
-------------------------
Significant outliers for TPR, FPR, and FDR calculations are based on q < 0.05. See Methods in the main paper.

*  `RDA1_propvar` proportion of variance explained by first RDA axis. RDA model: genotype ~ environment
*  `RDA2_propvar` proportion of variance explained by second RDA axis. RDA model: genotype ~ environment
*  `RDA1_propvar_corr` proportion of variance explained by first RDA axis. RDA model with structure correction: genotype ~ environment + Condition(PC1 + PC2)
*  `RDA2_propvar_corr` proportion of variance explained by second RDA axis. RDA model with structure correction: genotype ~ environment + Condition(PC1 + PC2)

*  `RDA1_temp_cor` output of `summary(rdaout)$biplot[2,1]`, which is the correlation between RDA1 and the temperature environmental variable. RDA model: genotype ~ environment
*  `RDA1_sal_cor`  output of `summary(rdaout)$biplot[1,1]`, which is the correlation between RDA1 and the salinity environmental variable. RDA model: genotype ~ environment
*  `RDA2_temp_cor` output of `summary(rdaout)$biplot[2,2]`, which is the correlation between RDA2 and the temperature environmental variable. RDA model: genotype ~ environment
*  `RDA2_sal_cor` output of `summary(rdaout)$biplot[1,2]`, which is the correlation between RDA2 and the salinity environmental variable. RDA model: genotype ~ environment

* `RDA_Va_temp_prop` proportion of additive genetic variance (Va) in the temperature trait explained by outliers in the RDA outlier analysis, following (Capblanq 2018, https://onlinelibrary.wiley.com/doi/abs/10.1111/1755-0998.12906). RDA model: genotype ~ environment
* `RDA_Va_temp_prop_corr` proportion of additive genetic variance (Va) in the temperature trait explained by outliers in the RDA outlier analysis, following (Capblanq 2018, https://onlinelibrary.wiley.com/doi/abs/10.1111/1755-0998.12906). RDA model with structure correction: genotype ~ environment + Condition(PC1 + PC2)
* `RDA_Va_sal_prop` proportion of additive genetic variance (Va) in the salinity trait explained by outliers in the RDA outlier analysis, following (Capblanq 2018, https://onlinelibrary.wiley.com/doi/abs/10.1111/1755-0998.12906). RDA model: genotype ~ environment
* `RDA_Va_sal_prop_corr` proportion of additive genetic variance (Va) in the salinity trait explained by outliers in the RDA outlier analysis, following (Capblanq 2018, https://onlinelibrary.wiley.com/doi/abs/10.1111/1755-0998.12906). RDA model with structure correction: genotype ~ environment + Condition(PC1 + PC2)

* `RDA_TPR` true positive rate of the RDA for all causal loci. Since RDA is a multidimensional analysis, I did not differentiate between loci that had causal effects on temperature or salinity. RDA model: genotype ~ environment
* `RDA_TPR_corr` true positive rate of the RDA for all causal loci. Since RDA is a multidimensional analysis, I did not differentiate between loci that had causal effects on temperature or salinity. RDA model with structure correction: genotype ~ environment + Condition(PC1 + PC2)

* `RDA_FDR_allSNPs` false discovery rate of the RDA outlier analysis based on the entire genome. RDA model: genotype ~ environment
* `RDA_FDR_allSNPs_corr` false discovery rate of the RDA outlier analysis based on the entire genome. RDA model with structure correction: genotype ~ environment + Condition(PC1 + PC2)

* `num_RDA_sig_causal` number of causal loci that are significant in the RDA analysis at q > 0.05. RDA model: genotype ~ environment
* `num_RDA_sig_neutral` number of neutral loci (LG 11-20) that are significant in the RDA analysis at q > 0.05. RDA model: genotype ~ environment
* `num_RDA_sig_causal_corr` number of causal loci that are significant in the RDA analysis at q > 0.05. RDA model with structure correction: genotype ~ environment + Condition(PC1 + PC2)
* `num_RDA_sig_neutral_corr` number of neutral loci (LG 11-20) that are significant in the RDA analysis at q > 0.05. RDA model with structure correction: genotype ~ environment + Condition(PC1 + PC2)

* `RDA_FDR_neutSNPs` an optimistic calculation of the false discovery rate of the RDA, including only causal loci and neutral loci not affected by selection (any non-causal loci that arises on the half of the genome affected by selection was excluded). RDA model: genotype ~ environment
* `RDA_FDR_neutSNPs_corr` an optimistic calculation of the false discovery rate of the RDA, including only causal loci and neutral loci not affected by selection (any non-causal loci that arises on the half of the genome affected by selection was excluded). RDA model with structure correction: genotype ~ environment + Condition(PC1 + PC2)

* `RDA_AUCPR_allSNPs`AUC-PR of the RDA outlier analysis based on the entire genome. RDA model: genotype ~ environment
* `RDA_AUCPR_neutSNPs` an optimistic calculation of the AUC-PR of the RDA, including only causal loci and neutral loci not affected by selection (any non-causal loci that arises on the half of the genome affected by selection was excluded). RDA model: genotype ~ environment
* `RDA_AUCPR_neutSNPs_corr` an optimistic calculation of the AUC-PR of the RDA, including only causal loci and neutral loci not affected by selection (any non-causal loci that arises on the half of the genome affected by selection was excluded).  RDA model with structure correction: genotype ~ environment + Condition(PC1 + PC2)

*  `RDA_FPR_neutSNPs` false positive rate of the RDA analysis, based only on neutral SNPs. RDA model: genotype ~ environment
*  `RDA_FPR_neutSNPs_corr` false positive rate of the RDA analysis, based only on neutral SNPs. RDA model with structure correction: genotype ~ environment + Condition(PC1 + PC2)

* `RDA_RDAmutpred_cor_tempEffect` # pearson's correlation between the predicted temperature effect from RDA and the true mutation effect on temperature. RDA model: genotype ~ environment
* `RDA_RDAmutpred_cor_salEffect` # pearson's correlation between the predicted salinity effect from RDA and the true mutation effect on salinity. RDA model: genotype ~ environment
* `RDA_absRDAmutpred_cor_tempVa` # pearson's correlation between the abs(predicted temperature effect from RDA) and the true mutation Va on temperature. RDA model: genotype ~ environment
* `RDA_absRDAmutpred_cor_salVa` # pearson's correlation between the abs(predicted salinity effect from RDA) and the true mutation Va on salinity. RDA model: genotype ~ environment

* `RDA_RDAmutpred_cor_tempEffect_structcorr` # pearson's correlation between the predicted temperature effect from RDA and the true mutation effect on temperature. RDA model: genotype ~ environment. RDA model with structure correction: genotype ~ environment + Condition(PC1 + PC2)
* `RDA_RDAmutpred_cor_salEffect_structcorr` # pearson's correlation between the predicted salinity effect from RDA and the true mutation effect on salinity. RDA model: genotype ~ environment. RDA model with structure correction: genotype ~ environment + Condition(PC1 + PC2)
* `RDA_absRDAmutpred_cor_tempVa_structcorr` # pearson's correlation between the abs(predicted temperature effect from RDA) and the true mutation Va on temperature. RDA model: genotype ~ environment. RDA model with structure correction: genotype ~ environment + Condition(PC1 + PC2)
* `RDA_absRDAmutpred_cor_salVa_structcorr` # pearson's correlation between the abs(predicted salinity effect from RDA) and the true mutation Va on salinity. RDA model: genotype ~ environment. RDA model with structure correction: genotype ~ environment + Condition(PC1 + PC2)

* `RDA_cor_RDA20000temppredict_tempPhen` correlation between the true temperature phenotype and that predicted from an RDA based on 20K SNPs. see `seed_Rout_RDA_predictions` for correlations with less loci used to make the prediction.  RDA model: genotype ~ environment
* `RDA_cor_RDA20000salpredict_salPhen` correlation between the true salinity phenotype and that predicted from an RDA based on 20K SNPs. see `seed_Rout_RDA_predictions`  for correlations with less loci used to make the prediction.  RDA model: genotype ~ environment
* `RDA_cor_RDA20000temppredict_tempPhen_structcorr` correlation between the true temperature phenotype and that predicted from an RDA based on 20K SNPs. see `seed_Rout_RDA_predictions` for correlations with less loci used to make the prediction. RDA model with structure correction: genotype ~ environment + Condition(PC1 + PC2)
* `RDA_cor_RDA20000salpredict_salPhen_structcorr` correlation between the true salinity phenotype and that predicted from an RDA based on 20K SNPs. see `seed_Rout_RDA_predictions`  for correlations with less loci used to make the prediction. RDA model with structure correction: genotype ~ environment + Condition(PC1 + PC2)


-------------------------
### Summary statistics of correlation between structure and environment
-------------------------

* `cor_PC1_temp` correlation between individual loading on PC1 from the principle components based on the Genotype-matrix (individual genotypes labeled as 0,1,2) and temperature of the deme where it was sampled
* `cor_PC1_sal` correlation between individual loading on PC1 from the principle components based on the Genotype-matrix and salnity of the deme where it was sampled
* `cor_PC2_temp` correlation between individual  loading on  PC2 from the principle components based on the Genotype-matrix and temperature of the deme where it was sampled
* `cor_PC2_sal` correlation between individual  loading on PC2 from the principle components based on the Genotype-matrix and salnity of the deme where it was sampled
* `cor_LFMMU1_temp` correlation between the individual loading on the latent factor 1 from the lfmm model based on temperature
* `cor_LFMMU1_sal` correlation between the individual loading on the latent factor 1 from the lfmm model based on salnity
* `cor_LFMMU2_temp` correlation between the individual loading on the latent factor 2 from the lfmm model based on temperature
* `cor_LFMMU2_sal` correlation between the individual loading on the latent factor 2 from the lfmm model based on salnity
* `cor_PC1_LFMMU1_temp` correlation between (individual loading on PC1 from the principle components based on the Genotype-matrix) and (individual loading on the latent factor 1 from the lfmm model based on temperature)
* `cor_PC1_LFMMU1_sal` correlation between (individual loading on PC1 from the principle components based on the Genotype-matrix) and (individual loading on the latent factor 1 from the lfmm model based on salinity)
* `cor_PC2_LFMMU1_temp` correlation between (individual loading on PC2 from the principle components based on the Genotype-matrix) and (individual loading on the latent factor 1 from the lfmm model based on temperature)
* `cor_PC2_LFMMU1_sal` correlation between (individual loading on PC2 from the principle components based on the Genotype-matrix) and (individual loading on the latent factor 1 from the lfmm model based on salinity)

-------------------------
### GWAS outputs
-------------------------
* `gwas_TPR_sal` true positive rate of the GWAS model for the salinity trait
* `gwas_TPR_temp` true positive rate of the GWAS model for the temperature trait
* `gwas_FDR_sal_neutbase` false discovery rate of the GWAS model for the salinity trait, only including QTNs and purely neutral loci unaffected by selection
* `gwas_FDR_temp_neutbase` false discovery rate of the GWAS model for the temperature trait, only including QTNs and purely neutral loci unaffected by selection
* `clinalparadigm_sal_proptop5GWASclines` proportion of the top 5% of GWAS loci with the smallest P-values for the salinity trait (true and false positives) that show clines
* `clinalparadigm_temp_proptop5GWASclines` proportion of the top 5% of GWAS loci with the smallest P-values for the temperature trait (true and false positives) that show clines
* `clinalparadigm_sal_propsigGWASclines` proportion of GWAS hits for the salinity trait (true and false positives) that show clines
* `clinalparadigm_temp_propsigGWASclines` proportion of GWAS hits for the temperature trait (true and false positives) that show clines