### `seed_Rout_simSummary.txt` (output after analysis of simulations) A table with information about the entire simulation

"seed" "n_samp_tot" "n_samp_per_pop" "sd_fitness_among_inds" "sd_fitness_among_pops" 
"final_LA" "K" "Bonf_alpha" "numCausalLowMAFsample" "all_corr_phen_temp" 
"subsamp_corr_phen_temp" "all_corr_phen_sal" "subsamp_corr_phen_sal" 
"num_causal_prefilter" "num_causal_postfilter" "num_non_causal" "num_neut_prefilter" 
"num_neut_postfilter" "num_neut_neutralgenome" "num_causal_temp" "num_causal_sal" 
"num_multiallelic" "meanFst" "va_temp_total" "va_sal_total" "Va_temp_sample" "Va_sal_sample" 
"nSNPs" "median_causal_temp_cor" "median_causal_sal_cor" "median_neut_temp_cor" 
"median_neut_sal_cor" "cor_VA_temp_prop" "cor_VA_sal_prop" "cor_TPR_temp" "cor_TPR_sal" 
"cor_FDR_allSNPs_temp" "cor_FDR_neutSNPs_temp" "cor_FDR_allSNPs_sal" "cor_FDR_neutSNPs_sal" 
"num_causal_sig_temp_corr" "num_causal_sig_sal_corr" "num_notCausal_sig_temp_corr" 
"num_notCausal_sig_sal_corr" "num_neut_sig_temp_corr" "num_neut_sig_sal_corr" 
"cor_AUCPR_temp_allSNPs" "cor_AUCPR_temp_neutSNPs" "cor_AUCPR_sal_allSNPs" "cor_AUCPR_sal_neutSNPs" 
"cor_af_temp_noutliers" "cor_af_sal_noutliers" "cor_FPR_temp_neutSNPs" "cor_FPR_sal_neutSNPs" 
"LEA3.2_lfmm2_Va_temp_prop" "LEA3.2_lfmm2_Va_sal_prop" "LEA3.2_lfmm2_TPR_temp" "LEA3.2_lfmm2_TPR_sal" 
"LEA3.2_lfmm2_FDR_allSNPs_temp" "LEA3.2_lfmm2_FDR_allSNPs_sal" "LEA3.2_lfmm2_FDR_neutSNPs_temp" 
"LEA3.2_lfmm2_FDR_neutSNPs_sal" "LEA3.2_lfmm2_AUCPR_temp_allSNPs" "LEA3.2_lfmm2_AUCPR_temp_neutSNPs" 
"LEA3.2_lfmm2_AUCPR_sal_allSNPs" "LEA3.2_lfmm2_AUCPR_sal_neutSNPs" "LEA3.2_lfmm2_mlog10P_tempenv_noutliers" 
"LEA3.2_lfmm2_mlog10P_salenv_noutliers" "LEA3.2_lfmm2_num_causal_sig_temp" "LEA3.2_lfmm2_num_neut_sig_temp" 
"LEA3.2_lfmm2_num_causal_sig_sal" "LEA3.2_lfmm2_num_causal_sig_temp.1" "LEA3.2_lfmm2_FPR_neutSNPs_temp" 
"LEA3.2_lfmm2_FPR_neutSNPs_sal" "RDA1_propvar" "RDA2_propvar" "RDA1_propvar_corr" "RDA2_propvar_corr" 
"RDA1_temp_cor" "RDA1_sal_cor" "RDA2_temp_cor" "RDA2_sal_cor" "RDA_Va_temp_prop" "RDA_Va_temp_prop_corr" 
"RDA_Va_sal_prop" "RDA_Va_sal_prop_corr" "RDA_TPR" "RDA_TPR_corr" "RDA_FDR_allSNPs" "RDA_FDR_allSNPs_corr" 
"num_RDA_sig_causal" "num_RDA_sig_neutral" "num_RDA_sig_causal_corr" "num_RDA_sig_neutral_corr" "RDA_FDR_neutSNPs" 
"RDA_FDR_neutSNPs_corr" "RDA_AUCPR_allSNPs" "RDA_AUCPR_neutSNPs" "RDA_AUCPR_neutSNPs_corr" "RDA_FPR_neutSNPs" 
"RDA_FPR_neutSNPs_corr" "RDA_RDAmutpred_cor_tempEffect" "RDA_RDAmutpred_cor_salEffect" "RDA_absRDAmutpred_cor_tempVa" 
"RDA_absRDAmutpred_cor_salVa" "RDA_RDAmutpred_cor_tempEffect_structcorr" "RDA_RDAmutpred_cor_salEffect_structcorr" 
"RDA_absRDAmutpred_cor_tempVa_structcorr" "RDA_absRDAmutpred_cor_salVa_structcorr" "RDA_cor_RDA20000temppredict_tempPhen" 
"RDA_cor_RDA20000salpredict_salPhen" "RDA_cor_RDA20000temppredict_tempPhen_structcorr" "RDA_cor_RDA20000salpredict_salPhen_structcorr" 
"cor_PC1_temp" "cor_PC1_sal" "cor_PC2_temp" "cor_PC2_sal" "cor_LFMMU1_temp" "cor_LFMMU1_sal" "cor_LFMMU2_temp" 
"cor_LFMMU2_sal" "cor_PC1_LFMMU1_temp" "cor_PC1_LFMMU1_sal" "cor_PC2_LFMMU1_temp" "cor_PC2_LFMMU1_sal"


After each simulation is analyzed in R, 

*`seed` simulation seed
* `n_samp_tot` total number of individuals sampled
* `nind_samp_per_pop` number of individuals sampled from each deme
* `sd_fitness_among_inds` variance in fitness among all sampled individuals in the simulation (sampling prob. is proportional to fitness to mimic viability selection)
* `sd_fitness_among_pops` variance in fitness among all demes in the simulation after sampling (sampling prob. is proportional to fitness to mimic viability selection)
* `final_LA` final amount of local adaptation in the simulation
* `K` number of populations used in analyses
* `Bonf_alpha` the significance threshold for P-values applied to the correlation
* `numCausalLowMAFsample` number of causal loci that were not filtered out, but were below the MAF cutoff. These were included in the calculations. **REVISIT**
* `all_corr_phen_temp` for all individuals, correlation between individual temp phenotype and environment temperature
* `subsamp_corr_phen_temp` after sampling 20 individuals from each patch with a probability based on their fitness, correlation individual temp phenotype and environment temperature
* `all_corr_phen_sal` for all individuals, correlation between individual sal phenotype and environment salinity
* `subsamp_corr_phen_sal` after sampling 20 individuals from each patch with a probability based on their fitness, correlation between individual sal phenotype and environment salinity
* `num_causal` number of causal loci in sim
* `num_non_causal`number of neutral loci in sim arising on the half of the genome where they could be linked to causal loci
* `num_neut` number of truly neutral loci in sim
* `num_causal_temp` number of loci with non-zero phenotypic effects on the temperature phenotype
* `num_causal_sal`number of loci with non-zero phenotypic effects on the salinity phenotype

* `va_temp_total` total additive genetic variance in the temperatuer trait, based on the entire sample
* `va_sal_total` total additive genetic variance in the salinity trait, based on the entire sample
* `mean_Fst` overall FST calculated from mean(T1)/mean(T2) in outflank
----

* `LEA3.2_lfmm2_Va_temp_prop` proportion of additive genetic variance (Va) in the temperature trait explained by outliers in the LFMM temp model
* `LEA3.2_lfmm2_Va_sal_prop` proportion of additive genetic variance (Va) in the saliniity trait explained by outliers in the LFMM salinity model
*  `LEA3.2_lfmm2_TPR_temp` true positive rate of the LFMM temp model for loci with alleles that have non-zero effects on the temperature phenotype
*  `LEA3.2_lfmm2_TPR_sal` true positive rate of the LFMM salinity model for loci with alleles that have non-zero effects on the salnity phenotype
*  `LEA3.2_lfmm2_FDR_allSNPs_temp` false discovery rate of the LFMM temp model for the entire genome      
* `LEA3.2_lfmm2_FDR_allSNPs_sal` false discovery rate of the LFMM temp model for the entire genome   
* `LEA3.2_lfmm2_FDR_neutSNPs_temp` an optimistic calculation of the false discovery rate of the LFMM temp model, including only causal loci and neutral loci not affected by selection (any non-causal loci that arises on the half of the genome affected by selection was excluded)
* `LEA3.2_lfmm2_FDR_neutSNPs_sal` an optimistic calculation of the false discovery rate of the LFMM salinity model, including only causal loci and neutral loci not affected by selection (any non-causal loci that arises on the half of the genome affected by selection was excluded)
*  `LEA3.2_lfmm2_AUCPR_temp_allSNPs` the AUC-PR of the lfmm temp model based on the entire genome  
* `LEA3.2_lfmm2_AUCPR_temp_neutSNPs` an optimistic calculation of the AUC-PR of the LFMM temp model, including only causal loci and neutral loci not affected by selection (any non-causal loci that arises on the half of the genome affected by selection was excluded)
* `LEA3.2_lfmm2_AUCPR_sal_allSNPs` the AUC-PR of the lfmm temp model based on the entire genome
* `LEA3.2_lfmm2_AUCPR_sal_neutSNPs` the AUC-PR of the lfmm salinity model based on the entire genome  

----

* `RDA_Va_temp_prop` proportion of additive genetic variance (Va) in the temperature trait explained by outliers in the RDA outlier analysis, following (Capblanq 2018, https://onlinelibrary.wiley.com/doi/abs/10.1111/1755-0998.12906)
* `RDA_Va_sal_prop` proportion of additive genetic variance (Va) in the salinity trait explained by outliers in the RDA outlier analysis, following (Capblanq 2018, https://onlinelibrary.wiley.com/doi/abs/10.1111/1755-0998.12906)
* `RDA_TPR` true positive rate of the RDA for all causal loci. Since RDA is a multidimensional analysis, I did not differentiate between loci that had causal effects on temperature or salinity.
* `RDA_FDR_allSNPs` false discovery rate of the RDA outlier analysis based on the entire genome
* `RDA_FDR_neutSNPs` an optimistic calculation of the false discovery rate of the RDA, including only causal loci and neutral loci not affected by selection (any non-causal loci that arises on the half of the genome affected by selection was excluded)
* `RDA_AUCPR_allSNPs`AUC-PR of the RDA outlier analysis based on the entire genome
* `RDA_AUCPR_neutSNPs` an optimistic calculation of the AUC-PR of the RDA, including only causal loci and neutral loci not affected by selection (any non-causal loci that arises on the half of the genome affected by selection was excluded)
* `cor_RDA20000_RDloadings_tempPhen` correlation between the true temperature phenotype and that predicted from an RDA based on 20K SNPs. see `seed_Rout_RDA_predictions`
* `cor_RDA20000_RDloadings_salPhen` correlation between the true salinity phenotype and that predicted from an RDA based on 20K SNPs. see `seed_Rout_RDA_predictions`  

----

* `cor_VA_temp_prop` proportion of VA in temperature phenotype explained by clinal outliers for temperature, based on kendall's correlation between deme allele frequency and deme temperature
* `cor_VA_sal_prop`  proportion of VA in salinity phenotype explained by clinal outliers for salinity, based on kendall's correlation between deme allele frequency and deme salinity
* `cor_TPR_temp` true positive rate for loci with non-zero effects on temperature, based on kendall's correlation between deme allele frequency and deme temperature 
* `cor_TPR_sal`true positive rate for loci with non-zero effects on salinity, based on kendall's correlation between deme allele frequency and deme salinity
* `cor_FDR_allSNPs_temp` false discovery rate of (kendall's correlation between deme allele frequency and deme temperature) for loci with non-zero effects on temperature 
* `cor_FDR_neutSNPs_temp` an optimistic calculation for false discovery rate of  (kendall's correlation between deme allele frequency and deme temperature) for loci with non-zero effects on temperature, excluding non-causal loci in half of genome affected by selection 
* `cor_FDR_allSNPs_sal` false discovery rate of (kendall's correlation between deme allele frequency and deme salinity) for loci with non-zero effects on salinity
* `cor_FDR_neutSNPs_sal` an optimistic calculation for false discovery rate of  (kendall's correlation between deme allele frequency and deme temperature) for loci with non-zero effects on temperature, excluding non-causal loci in half of genome affected by selection
* `cor_AUCPR_temp_allSNPs` AUC-PR of kendall's correlation between deme allele frequency and deme temperature, based on the whole genome and causal loci for temperature
* `cor_AUCPR_temp_neutSNPs` an optimistic estimate of AUC-PR of kendall's correlation between deme allele frequency and deme temperature, based on causal loci for temperature and neutral loci not affected by selection (excluding non-causal loci in half of genome affected by selection)
* `cor_AUCPR_sal_allSNPs` AUC-PR of kendall's correlation between deme allele frequency and deme salinity, based on the whole genome and causal loci for salinity
* `cor_AUCPR_sal_neutSNPs` an optimistic estimate of AUC-PR of kendall's correlation between deme allele frequency and deme salinity, based on causal loci for salinity and neutral loci not affected by selection (excluding non-causal loci in half of genome affected by selection)

----

* `median_causal_temp_cor` median abs(Spearman's correlation) between allele frequency and temperature for causal loci
* `median_causal_sal_cor` median abs(Spearman's correlation)  between allele frequency and salinity for causal loci
* `median_neut_temp_cor` median abs(Spearman's correlation)  between allele frequency and temperature for neutral loci
* `median_neut_sal_cor` median abs(Spearman's correlation)  between allele frequency and salinity for neutral loci

----

* `cor_PC1_temp` correlation between individual loading on PC1 from the principle components based on the Genotype-matrix (individual genotypes labeled as 0,1,2) and temperature of the deme where it was sampled
* `cor_PC1_sal` correlation between individual loading on PC1 from the principle components based on the Genotype-matrix and salnity of the deme where it was sampled
* `cor_PC2_temp` correlation between individual  loading on  PC2 from the principle components based on the Genotype-matrix and temperature of the deme where it was sampled
* `cor_PC2_sal` correlation between individual  loading on PC2 from the principle components based on the Genotype-matrix and salnity of the deme where it was sampled
* `cor_LFMMU1_temp` correlation between the individual loading on the latent factor 1 from the lfmm model based on temperature
* `cor_LFMMU1_sal` correlation between the individual loading on the latent factor 1 from the lfmm model based on salnity
* `cor_LFMMU2_temp` correlation between the individual loading on the latent factor 2 from the lfmm model based on temperature
* `cor_LFMMU2_sal` correlation between the individual loading on the latent factor 2 from the lfmm model based on salnity
* `cor_PC1_LFMMU1_temp` correlation between (individual loading on PC1 from the principle components based on the Genotype-matrix) and (individual loading on the latent factor 1 from the lfmm model based on temperature)
* `cor_PC1_LFMMU1_sal` correlation between (individual loading on PC1 from the principle components based on the Genotype-matrix) and (individual loading on the latent factor 1 from the lfmm model based on salinity)
* `cor_PC2_LFMMU1_temp` correlation between (individual loading on PC2 from the principle components based on the Genotype-matrix) and (individual loading on the latent factor 1 from the lfmm model based on temperature)
* `cor_PC2_LFMMU1_sal` correlation between (individual loading on PC2 from the principle components based on the Genotype-matrix) and (individual loading on the latent factor 1 from the lfmm model based on salinity)

----

* `cor_af_temp_noutliers` number of outliers for cor(af,temp) after Bonferroni correction
* `cor_af_sal_noutliers` number of outliers for cor(af,salinity) after Bonferroni correction
* `nSNPs` total number of SNPs in analysis
* `cor_FPR_temp_neutSNPs` false positive rate in cor(af,temp) after Bonferroni correction, based on loci unaffected by selection
* `cor_FPR_sal_neutSNPs` false positive rate in cor(af,sal) after Bonferroni correction, based on loci unaffected by selection
* `LEA3.2_lfmm2_mlog10P_tempenv_noutliers` number of outliers for the lfmm temp model (qvalue <0.05)
* `LEA3.2_lfmm2_mlog10P_salenv_noutliers` number of outliers for the lfmm salinity model (qvalue <0.05)
* `LEA3.2_lfmm2_num_causal_sig_temp` number of causal loci on the temp trait, significant in the lfmm temp model (qvalue <0.05)
* `LEA3.2_lfmm2_num_neut_sig_temp` number of neutral loci false positives (only neutral loci not affected by selection), significant in the lfmm temp model (qvalue <0.05)
* `LEA3.2_lfmm2_num_causal_sig_sal` number of causal loci on the salinity trait, significant in the lfmm salinity model (qvalue <0.05)
* `LEA3.2_lfmm2_num_neut_sig_sal` number of neutral loci false positives (only neutral loci not affected by selection), significant in the lfmm salinity model (qvalue <0.05)
*  `LEA3.2_lfmm2_FPR_neutSNPs_temp` false positive rate of lfmm temperature model
*  `LEA3.2_lfmm2_FPR_neutSNPs_sal` false positive rate of lfmm salinity model

----

*  `RDA1_propvar` proportion of variance explained by first RDA axis
*  `RDA2_propvar` proportion of variance explained by second RDA axis
*  `RDA1_temp_cor` output of `summary(rdaout)$biplot[2,1]`, which is the correlation between RDA1 and the temperature environmental variable
*  `RDA1_sal_cor`  output of `summary(rdaout)$biplot[1,1]`, which is the correlation between RDA1 and the salinity environmental variable
*  `RDA2_temp_cor` output of `summary(rdaout)$biplot[2,2]`, which is the correlation between RDA2 and the temperature environmental variable
*  `RDA2_sal_cor` output of `summary(rdaout)$biplot[1,2]`, which is the correlation between RDA2 and the salinity environmental variable
*  `RDA_mlog10P_sig_noutliers` number of outliers in the RDA analysis 
*  `RDA_FPR_neutSNPs` false positive rate of the RDA analysis, based only on neutral SNPs
* `RDA_RDAmutpred_cor_tempEffect` # pearson's correlation between the predicted temperature effect from RDA and the true mutation effect on temperature
* `RDA_RDAmutpred_cor_salEffect` # pearson's correlation between the predicted salinity effect from RDA and the true mutation effect on salinity
* `RDA_absRDAmutpred_cor_tempVa` # pearson's correlation between the abs(predicted temperature effect from RDA) and the true mutation Va on temperature
* `RDA_absRDAmutpred_cor_salVa` # pearson's correlation between the abs(predicted salinity effect from RDA) and the true mutation Va on salinity

